src/ast.ts

```diff
@@ -1131,6 +1161,18 @@ export class FunctionTypeNode extends TypeNode {
   returnType: TypeNode;
   /** Explicitly provided this type, if any. */
   explicitThisType: NamedTypeNode | null; // can't be a function
+
+  equals(node: FunctionTypeNode): bool {
+    if (!TypeNode.arrayEquals(this.parameterTypes, node.parameterTypes)) return false;
+    if (!this.returnType.equals(node.returnType)) return false;
+    if ((this.explicitThisType == null && node.explicitThisType == null)) return true;
+    if (this.explicitThisType == null || node.explicitThisType == null) return false;
+    return true;
+  }
+
+  get parameterTypes(): TypeNode[] {
+    return this.parameters.map((param: ParameterNode): TypeNode => param.type);
```

Comment:
I know this is a draft currently, but please avoid `.map` and similar in the compiler for now.

Reply:
I agree. In any case first class functional methods much slower than vanilla loops in JS/TS. On AssemblyScript (later when it will self-hosted) it also will be slower while we don't learn how replace/inline such high ordered routines to plain loops

Reply:
Ah shoot it's used a good bit.  I added a map function for iterables too.  Is this a concern of style or compatibility? If the latter, could I use this function since arrays are iterable?

Reply:
I've tried to avoid these because it'd compile to an anonymous function currently that must be called indirectly in a loop, so a plain loop is much more efficient. Also suggesting to avoid iterators. The compiler only uses `for ... of` as a last resort where there's no alternative.

Reply:
Ah okay. It seems something that would optimized by the VM's JIT, but I'll remove them.

Reply:
Yeah, this is more in preparation of compiling the compiler to Wasm eventually. The less surface area we have where something can go wrong, the less frustrating this process will be.

Reply:
@willemneal v8 JIT can't optimize such sophisticated cases unfortunately. See this [bench](https://gist.github.com/MaxGraey/e7486833ae3a7abe072ac35afce8a240)

Reply:
Ah shoot oh well (perhaps we can take advantage of tail calls in wasm later).